modules/standard/FileSystem.chpl

```diff
@@ -109,6 +109,198 @@ proc chown(name: string, uid: int, gid: int) {
+   message.  If metadata is set to true, will also copy the metadata of the
+   file to be copied.  May halt with other error messages.
+   src: the source file whose contents and permissions are to be copied
+   dest: the destination file for the contents and permissions.
+   metadata: a boolean indicating whether to copy metadata (uid, gid, time of
+             last access and time of modification) associated with the source
+             file.
+*/
+proc copy(src: string, dest: string, metadata: bool = false) {
+  var err: syserr = ENOERR;
+  copy(err, src, dest, metadata);
+  if err != ENOERR then ioerror(err, "in copy(" + src + ", " + dest + ")");
+}
+
+/* Copies the contents of the file indicated by src into the file indicated
+   by dest, replacing dest if it already exists (and is different than src).
+   If the dest is not writable or src and dest are the same file, generates
+   an error.  Does not copy metadata.
+   err: a syserr used to indicate if an error occurred
+   src: the source file whose contents are to be copied.
+   dest: the destination of the contents.
+*/
+proc copyFile(out err: syserr, src: string, dest: string) {
+  // This implementation is based off of the python implementation for copyfile,
+  // with some slight differences.  That implementation was found at:
+  // https://bitbucket.org/mirror/cpython/src/c8ce5bca0fcda4307f7ac5d69103ce128a562705/Lib/shutil.py?at=default
+  // I did not look at the other functions in that file, except for copyfileobj
+  // (which copyfile called).
+  if (!exists(src)) {
+    err = ENOENT;
+    // Source didn't exist, we can't copy it.
+    return;
+  }
+  if (isDir(err, src) || isDir(err, dest)) {
+    // If the source is a directory, the user has made a mistake, so return an
+    // error.  The same is true if the destination is a directory.
+    err = EISDIR;
+    return;
+  }
+
+  if (err == ENOENT) {
+    err = ENOERR;
+    // We don't care if dest did not exist before, we'll create or overwrite it
+    // anyways.  We already know src exists.
+  } else if (sameFile(err, src, dest)) {
+    // Check if the files are the same, error if yes
+
+    // Don't need to check if they're the same file when we know dest didn't
+    // exist.
+    err = EINVAL;
+    return;
+    // The second argument is invalid if the two arguments are the same.
+  }
+
+  // Lydia note (03/04/2014): These enclosing curly braces are to avoid the bug
+  // found in test/io/lydia/outArgEarlyExit*.future.  When those futures are
+  // resolved, the braces should be removed.
+  {
+    // Open src for reading, open dest for writing
+    var srcFile = open(src, iomode.r);
+    var destFile = open(dest, iomode.cw);
+    var srcChnl = srcFile.reader(kind=ionative, locking=false);
+    var destChnl = destFile.writer(kind=ionative, locking=false);
+    // read in, write out.
+    var line: [0..1023] uint(8);
+    var numRead: int = 0;
+    while (srcChnl.readline(line, numRead=numRead, error=err)) {
```

Comment:
Ncurses!
One day I hope to have a QIO routine that just does "copy the rest of this channel to that channel" so that normal people don't have to think so much :)